osdconfig/interface.go

```diff
@@ -0,0 +1,63 @@
+// osdconfig is a package to work with distributed config parameters
+package osdconfig
+
+import "context"
+
+// A config manager interface allows management of osdconfig parameters
+// It defines setters, getters and callback management functions
+type ConfigManager interface {
+	// GetClusterConf fetches cluster configuration data from a backend such as kvdb
+	GetClusterConf() (*ClusterConfig, error)
+
+	// Fetch node configuration data using node id
+	GetNodeConf(nodeID string) (*NodeConfig, error)
+
+	// SetClusterConf pushes cluster configuration data to the backend
+	// It is assumed that the backend will notify the implementor of this interface
+	// when a change is triggered
+	SetClusterConf(config *ClusterConfig) error
+
+	// SetNodeConf pushes node configuration data to the backend
+	// It is assumed that the backend will notify the implementor of this interface
+	// when a change is triggered
+	SetNodeConf(config *NodeConfig) error
+
+	// Register callback functions that will get triggered on a change to either
+	// cluster configuration or node configuration data in the backend.
+	// The callback function returns two channels: A channel to write to and a
+	// channel to read on. Context is provided to the callback function in order
+	// to manage cleanup during context cancellations
+	Register(name string, watcherType Watcher,
+		opt interface{},
+		cb func(ctx context.Context, opt interface{}) (chan<- *DataWrite,
+			<-chan *DataRead)) error
+
+	// Run executes callback functions, however, it is expected that users never
+	// call this function directly. Run is called behind the scenes using scheduler
+	Run(wd *DataWrite)
```

Comment:
this should not be exposed to clients, should be removed from the interface

Reply:
I agree as you can tell with the comments I wrote. It is, however, very useful for tests and several of my tests currently depend on it.

Reply:
imho test should not expose implementation details in interface.
if you want to test implementation, use casts.

Reply:
ok, let me take a look. btw, this will also get exposed through REST, where I don't plan to expose most of these functions. For devs, including myself, I anticipate needing these down the lane though.